google/cloud/testing_util/validate_metadata.h

```diff
@@ -15,20 +15,60 @@
 #ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_TESTING_UTIL_VALIDATE_METADATA_H
 #define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_TESTING_UTIL_VALIDATE_METADATA_H
 
-#include "google/cloud/status.h"
+#include "google/cloud/internal/absl_str_replace_quiet.h"
 #include "google/cloud/version.h"
 #include "absl/types/optional.h"
+#include <google/api/annotations.pb.h>
+#include <google/protobuf/descriptor.h>
 #include <gmock/gmock.h>
 #include <grpcpp/generic/async_generic_service.h>
 #include <grpcpp/grpcpp.h>
 #include <map>
+#include <regex>
 #include <string>
 
 namespace google {
 namespace cloud {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace testing_util {
 
+/**
+ * We do not use `::testing::MatchesRegex` because our Windows builds use a
+ * googletest built with `GTEST_USES_SIMPLE_RE`, instead of
+ * `GTEST_USES_POSIX_RE`.
+ */
+MATCHER_P(MatchesMetadataRegex, pattern,
+          "matches the pattern: \"" + pattern + "\"") {
+  // Translate the pattern into a form that can be used by std::regex
+  std::regex regex(absl::StrReplaceAll(pattern, {{"*", "[^/]+"}}));
+  return std::regex_match(arg, regex);
+}
+
+std::map<std::string, std::string> ExtractMDFromHeader(std::string header);
+
+std::map<std::string, std::string> FromHttpRule(
+    google::api::HttpRule const& http,
+    absl::optional<std::string> const& resource_name);
+
+template <typename Request>
+std::map<std::string, std::string> ExtractRoutingHeaders(
+    std::string const& method, Request const&,
```

Comment:
Consider consuming a `google::protobuf::Message const& request` instead and keeping in the .cc file.

Reply:
Done.